core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala

```diff
@@ -258,6 +258,16 @@ class PaigesScalacheckTest extends FunSuite {
     forAll { (a: Doc) => assert(okay(a)) }
   }
 
+  test("Union(a1, b1) + Union(a2, b2) === Union(a1 + a2, b1 + b2)") {
+    import Doc._
+    forAll { (d1: Doc, d2: Doc) =>
+      (d1.grouped, d2.grouped) match {
+        case (d1@Union(a1, b1), d2@Union(a2, b2)) => Concat(d1, d2) === Union(Concat(a1, a2), Concat(b1, b2))
```

Comment:
I would think concat is more like `*`

so, I think this is `(a1 + b1)*(a2 +b2) = (a1 a2 + a1 b2 + b1 a2 + b1 b2)`

I think this is passing either because we didn't look hard enough for counter examples, or because we we tend to not have a decision in the second part of the union: the first choice determines what the second choice will do usually.

But I think we can simplify the law:

`(a + b)*c == ac + bc` and `a*(b + c) = ab + ab` and if we do those two laws, the current one should follow.

What do you think about writing those two laws instead?

Reply:
```
(a1 + b1)*(a2 +b2) = (a1 a2 + a1 b2 + b1 a2 + b1 b2)
```

I don't think this is correct.  In `a + b` we have invariants about the shape of `a`.  "+", which I assume you're using for `Union` is not associative, since (a + b) + c doesn't satisfy even the most basic invariants.  So in `a1 a2 + (a1 b2 + (b1 a2 + b1 b2))` which would be the only way to parse that, you have `a1 b2` on the left of a Union, which in general breaks invariants.  The only parts of that sum that make sense together are `a1 a2` and `b1 b2`.  Thus I think that the rule as written in the tests is correct.

Reply:
```
(a + b)*c == ac + bc and a*(b + c) = ab + ab
```

These make sense when `ac` is flat, which is equivalent to `c` being flat, since `a` is flat by invariant.  That's a nice equality.  Same for the other, so yes, I like those rules.  I'll add them.  They don't imply the former rule due to the restriction on `a`, (`c` respectively) being flat.

Reply:
I think we are getting lost here...

There are things we know to be true which we use for efficiency, but others that are true for the API to make sense.

So, do you disagree that `(a + b)*c == (ac + bc)`? if not, why don't we add that and the other `c * (a + b) = ca + cb`?

if those two laws are true, I think the other law works (I think with any version of the parens actually).

If I am wrong, we should see some violations here?

If not, them my claim about us not using enough trials is likely true since two conflicting rules can't both be true.

Reply:
I mean:
```scala
 forAll { (d1: Doc, d2: Doc) =>
      d1.grouped match {
        case d1@Union(a1, b1) => 
          assert(Concat(d1, d2) === Union(Concat(a1, d2), Concat(b1, d2)))
        case _ => ()
  }
```
and the opposite variant.

Reply:
Ah, you're right that there should be 4 options for a concat of unions, so this test can't be right.  Thanks. In general I don't think it's reducible anymore.  In the special case that one side is flat it works out.  I'll add those tests for your proposals in another PR.  







, but I'm struggling to understand why my